src/core.js

```diff
@@ -63,53 +71,106 @@ function createModule( name, testEnvironment ) {
 	return module;
 }
 
-extend( QUnit, {
-	on,
+function processModule( { name, testEnvironment, executeNow, skip = false, todo = false } ) {
+	let module = createModule( { name, testEnvironment, skip, todo } );
 
-	// Call on start of module test to prepend name to all tests
-	module: function( name, testEnvironment, executeNow ) {
-		if ( arguments.length === 2 ) {
-			if ( objectType( testEnvironment ) === "function" ) {
-				executeNow = testEnvironment;
-				testEnvironment = undefined;
-			}
+	// Move any hooks to a 'hooks' object
+	if ( module.testEnvironment ) {
+		module.hooks = {
+			before: module.testEnvironment.before,
+			beforeEach: module.testEnvironment.beforeEach,
+			afterEach: module.testEnvironment.afterEach,
+			after: module.testEnvironment.after
+		};
+
+		delete module.testEnvironment.before;
+		delete module.testEnvironment.beforeEach;
+		delete module.testEnvironment.afterEach;
+		delete module.testEnvironment.after;
+	}
+
+	const moduleFns = {
+		before: setHook( module, "before" ),
+		beforeEach: setHook( module, "beforeEach" ),
+		afterEach: setHook( module, "afterEach" ),
+		after: setHook( module, "after" )
+	};
+
+	const currentModule = config.currentModule;
+	if ( objectType( executeNow ) === "function" ) {
+		moduleStack.push( module );
+		config.currentModule = module;
+		executeNow.call( module.testEnvironment, moduleFns );
+		moduleStack.pop();
+		module = module.parentModule || currentModule;
+	}
+
+	config.currentModule = module;
+}
+
+function module( name, testEnvironment, executeNow ) {
+	if ( focused ) {
+		return;
+	}
+
+	if ( arguments.length === 2 ) {
+		if ( objectType( testEnvironment ) === "function" ) {
+			executeNow = testEnvironment;
+			testEnvironment = undefined;
 		}
+	}
+
+	processModule( { name, testEnvironment, executeNow } );
+}
 
-		let module = createModule( name, testEnvironment );
-
-		// Move any hooks to a 'hooks' object
-		if ( module.testEnvironment ) {
-			module.hooks = {
-				before: module.testEnvironment.before,
-				beforeEach: module.testEnvironment.beforeEach,
-				afterEach: module.testEnvironment.afterEach,
-				after: module.testEnvironment.after
-			};
-
-			delete module.testEnvironment.before;
-			delete module.testEnvironment.beforeEach;
-			delete module.testEnvironment.afterEach;
-			delete module.testEnvironment.after;
+module.only = function() {
+	if ( focused ) {
+		return;
+	}
+
+	moduleStack.length = 0;
```

Comment:
I don't think we should do this. If you focus a module that is nested within another, then the `moduleStack` would still need to be used.